### PR TITLE
Add x86 intrinsics support for sha1 and sha2

### PR DESCRIPTION
**Pull request by newpavlov**
Written using #90 and the software implementation as a reference.

This PR also simplifies handling of ARM backend and deprecates `asm-aarch64` feature.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
@newpavlov I agree, your code looks much nicer :) 

Some thoughts for my specific use cases

(1) Bringing back dynamic detection would be great, often times the code will be built once and then run on different cpus in my case. 
(2) ~~now that `input_blocks` is a thing again it will be more efficient to process multiple blocks at once, like I am doing here: https://github.com/filecoin-project/rust-fil-proofs/blob/master/sha2raw/src/sha256_intrinsics.rs#L12~~ Looks like you already did this
(3) currently all implementations do have some overhead and additional padding in my measurements, which is why I ended up implementing this "raw" construction https://github.com/filecoin-project/rust-fil-proofs/blob/master/sha2raw/src/sha256.rs#L25 which avoids going through the `block_buffer` and can only process appropriately sized and pre padded chunks. It would be great to find a way such that passing in slices of matching data could have the same speed.

**Comment by Fleshgrinder** [comment]
I second the dynamic detection request. This is important for software that is installed on many different systems.

**Comment by newpavlov** [comment]
>Bringing back dynamic detection would be great

The code is written with dynamic dispatch in mind, I just need to write CPUID detection using `getrandom` code as a reference (probably I will move it into a separate crate).

>currently all implementations do have some overhead and additional padding in my measurements

If you don't need padding, then I think the easiest way will be to use the compression functions directly, which can be accessed after enabling the `compress` feature. I guess we also could add imports for initialization constants.

**Comment by newpavlov** [comment]
Runtime detection is back! If you interested, it's powered by the [`cpuid-bool`](https://github.com/RustCrypto/utils/blob/master/cpuid-bool/src/lib.rs) crate.

@linkmauve
Can you verify that crates are not broken on ARM?

**Review by tarcieri — APPROVED** [review]

**Review by dignifiedquire — APPROVED** [review]
very nice, thank you!

File: sha1/src/compress/x86.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    state[1] = _mm_extract_epi32(state_abcd, 2) as u32;
+    state[2] = _mm_extract_epi32(state_abcd, 1) as u32;
+    state[3] = _mm_extract_epi32(state_abcd, 0) as u32;
+    state[4] = _mm_extract_epi32(state_e, 3) as u32;
+}
+
+pub fn compress(state: &mut [u32; 5], blocks: &[[u8; 64]]) {
+    // TODO: Replace with https://github.com/rust-lang/rfcs/pull/2725
+    // after stabilization
+    if cpuid_bool::cpuid_bool!("sha", "sse2", "ssse3", "sse4.1") {
```
Comment (jethrogb) [inline sha1/src/compress/x86.rs:101]: This code breaks platforms that don't support CPUID (e.g. SGX). You should be using https://doc.rust-lang.org/std/macro.is_x86_feature_detected.html, which is the standard way all platforms support feature detection.

Comment (newpavlov) [reply sha1/src/compress/x86.rs:101]: Note that `is_x86_feature_detected` is not available in `libcore` and the linked PR proposes changes to change it. See #183 for discussion as how to fix SGX.

**Pull request merged**